pkg/instancetype/webhooks/vm/admitter.go

```diff
@@ -0,0 +1,120 @@
+ *
+ */
+package vm
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sfield "k8s.io/apimachinery/pkg/util/validation/field"
+	"kubevirt.io/client-go/kubecli"
+
+	"kubevirt.io/kubevirt/pkg/instancetype/apply"
+	"kubevirt.io/kubevirt/pkg/instancetype/conflict"
+	"kubevirt.io/kubevirt/pkg/instancetype/find"
+	preferenceFind "kubevirt.io/kubevirt/pkg/instancetype/preference/find"
+	"kubevirt.io/kubevirt/pkg/instancetype/preference/requirements"
+	"kubevirt.io/kubevirt/pkg/instancetype/preference/validation"
+
+	virtv1 "kubevirt.io/api/core/v1"
+	"kubevirt.io/api/instancetype/v1beta1"
+)
+
+type instancetypeFinder interface {
+	Find(*virtv1.VirtualMachine) (*v1beta1.VirtualMachineInstancetypeSpec, error)
+}
+
+type preferenceFinder interface {
+	FindPreference(*virtv1.VirtualMachine) (*v1beta1.VirtualMachinePreferenceSpec, error)
+}
+
+type requirementsChecker interface {
+	Check(*v1beta1.VirtualMachineInstancetypeSpec,
+		*v1beta1.VirtualMachinePreferenceSpec,
+		*virtv1.VirtualMachineInstanceSpec,
+	) (conflict.Conflicts, error)
+}
+
+type applyVMIHandler interface {
+	ApplyToVMI(
+		*k8sfield.Path,
+		*v1beta1.VirtualMachineInstancetypeSpec,
+		*v1beta1.VirtualMachinePreferenceSpec,
+		*virtv1.VirtualMachineInstanceSpec,
+		*metav1.ObjectMeta,
+	) conflict.Conflicts
+}
+
+type admitter struct {
+	instancetypeFinder
+	preferenceFinder
+	applyVMIHandler
+	requirementsChecker
+}
+
+func NewAdmitter(virtClient kubecli.KubevirtClient) *admitter {
+	return &admitter{
+		instancetypeFinder:  find.NewSpecFinder(nil, nil, nil, virtClient),
+		preferenceFinder:    preferenceFind.NewSpecFinder(nil, nil, nil, virtClient),
+		requirementsChecker: requirements.New(),
+		applyVMIHandler:     apply.NewVMIApplier(),
+	}
+}
+
+func (a *admitter) ApplyToVM(vm *virtv1.VirtualMachine) (
+	*v1beta1.VirtualMachineInstancetypeSpec,
+	*v1beta1.VirtualMachinePreferenceSpec,
+	[]metav1.StatusCause,
+) {
+	instancetypeSpec, err := a.Find(vm)
+	if err != nil {
+		return nil, nil, []metav1.StatusCause{{
+			Type:    metav1.CauseTypeFieldValueNotFound,
+			Message: fmt.Sprintf("Failure to find instancetype: %v", err),
+			Field:   k8sfield.NewPath("spec", "instancetype").String(),
+		}}
+	}
+
+	preferenceSpec, err := a.FindPreference(vm)
+	if err != nil {
+		return nil, nil, []metav1.StatusCause{{
+			Type:    metav1.CauseTypeFieldValueNotFound,
+			Message: fmt.Sprintf("Failure to find preference: %v", err),
+			Field:   k8sfield.NewPath("spec", "preference").String(),
+		}}
+	}
+
+	if spreadConflict := validation.CheckSpreadCPUTopology(instancetypeSpec, preferenceSpec); spreadConflict != nil {
+		return nil, nil, spreadConflict.StatusCauses()
+	}
+
+	conflicts := a.ApplyToVMI(
+		k8sfield.NewPath("spec", "template", "spec"),
+		instancetypeSpec,
+		preferenceSpec,
+		&vm.Spec.Template.Spec,
+		&vm.Spec.Template.ObjectMeta,
+	)
+
+	if len(conflicts) == 0 {
+		return instancetypeSpec, preferenceSpec, nil
+	}
+
+	return nil, nil, conflicts.StatusCauses()
```

Comment:
What do you think about swapping the success/failure cases? It looks unusual like this to me.